Implements #3069 - allows use of pre-hashed passwords with userpass backend

Ivan-Roger opened the pull request:
This is a cherry-pick of changes suggested in hashicorp/vault#3225.

EventideGlow committed:
Implements #3069 - allows use of pre-hashed passwords with userpass backend

hashicorp-cla commented:
[![CLA assistant check](https://cla.hashicorp.com/pull/badge/not_signed)](https://cla.hashicorp.com/hashicorp/vault?pullRequest=24651)


Thank you for your submission! We require that all contributors sign our Contributor License Agreement ("CLA") before we can accept the contribution. [Read and sign the agreement](https://cla.hashicorp.com/hashicorp/vault?pullRequest=24651)

[Learn more about why HashiCorp requires a CLA and what the CLA includes](https://www.hashicorp.com/cla)


---
**2** out of **3** committers have signed the CLA.
- [x] peteski22
- [x] Ivan-Roger
- [ ] EventideGlow

<sub>Have you signed the CLA already but the status is still pending? [Recheck it](https://cla.hashicorp.com/check/hashicorp/vault?pullRequest=24651).</sub>

Ivan-Roger committed:
Fix issues from original PR

Improve errors messages to mention possibly missing password_hash.
Simplify checks for provided password during update.

schavis commented on website/content/api-docs/auth/userpass.mdx:
```diff
@@ -27,8 +27,8 @@ Create a new user or update an existing user. This path honors the distinction b
-- `password` `(string: <required>)` - The password for the user. Only required
-  when creating the user.
+- `password` `(string: <required if password_hash is not given>)` - The password for the user. Only required when creating the user.
+- `password_hash` `(string: <required if password is not given>)` - Pre-hashed version of this user's password in bcrypt format. Mutually exclusive with `password`.
```
```suggestion
- `password_hash` `(string: <required if password is not given>)` - Pre-hashed password for the current user in bcrypt format. Mutually exclusive with `password`.
```

schavis commented on website/content/api-docs/auth/userpass.mdx:
```diff
@@ -27,8 +27,8 @@ Create a new user or update an existing user. This path honors the distinction b
 - `username` `(string: <required>)` – The username for the user. Accepted characters: alphanumeric plus "_", "-", "." (underscore, hyphen and period); username cannot begin with a hyphen, nor can it begin or end with a period.
-- `password` `(string: <required>)` - The password for the user. Only required
-  when creating the user.
+- `password` `(string: <required if password_hash is not given>)` - The password for the user. Only required when creating the user.
```
```suggestion
- `password` `(string: <required if password_hash is not given>)` - Password for the current user. Only required when creating the user.  Mutually exclusive with `password_hash`.
```

updated to match the language in `password_hash`

Ivan-Roger committed:
Apply suggestions from code review

Co-authored-by: Sarah Chavis <62406755+schavis@users.noreply.github.com>

schavis commented:
Content changes lgtm. Please make sure to sign the CLA so your changes can be merged once someone from the Vault ENG team signs off on the code changes

Ivan-Roger commented:
@schavis Unfortunately the missing CLA signature comes from the original author (I cherry-picked their commit). He does not seem to be active anymore. Would it be possible to bypass his signature ?

Ivan-Roger committed:
Merge branch 'main' into feat/auth/userpass-with-hashed-password

peteski22 committed:
Merge branch 'main' into feat/auth/userpass-with-hashed-password

peteski22 commented:
Hi @Ivan-Roger thanks for trying to get this over the line, as the original contributor hasn't signed the CLA, I'm going to move this on via a different method.

I will close your PR in favour of https://github.com/hashicorp/vault/pull/25862
